File: reference/internal/metrics/aggregate.go
// Package metrics holds pure aggregation logic for the metrics tab.
//
// It depends only on internal/azdevops types and has no I/O or UI, so the
// core can be exercised entirely by table-driven tests.
package metrics

import (
	"sort"
	"time"

	"github.com/Elpulgo/azdo/internal/azdevops"
)

// Thresholds carries the configurable cut-offs that drive flagging
// and the overloaded marker on the per-user table.
type Thresholds struct {
	ActiveStaleDays int         // dwell in Active longer than this -> active-stale flag
	RFTStaleDays    int         // dwell in Ready for Test longer than this -> rft-stale flag
	WIPLimit        int         // strictly more in-flight items than this -> Overloaded
	States          StateConfig // canonical state names; matched case-insensitively
}

// UserMetrics is one row of the per-developer table.
type UserMetrics struct {
	User         string
	InFlight     int           // Active + Ready for Test
	ActiveCount  int
	RFTCount     int
	PointsClosed float64       // story points on items closed inside the interval
	OldestActive time.Duration // worst current-state dwell among Active items
	OldestRFT    time.Duration // worst current-state dwell among RFT items
	Stalled      int           // count of this user's items that tripped a threshold
	Overloaded   bool          // InFlight > Thresholds.WIPLimit
}

// ItemFlag is one row of the worst-first stuck-items digest.
type ItemFlag struct {
	ID      int
	Title   string
	Project string
	User    string
	State   string
	Dwell   time.Duration
	Reason  string // "active-stale" or "rft-stale"
}

const (
	reasonActiveStale = "active-stale"
	reasonRFTStale    = "rft-stale"
)

// Aggregate rolls a flat slice of work items into per-developer rows and a
// flat list of flagged items, applying `th` thresholds. Items in states other
// than Active / Ready for Test / Closed are ignored.
//
// Closed items contribute to PointsClosed only when their ClosedDate falls
// strictly after intervalStart.
func Aggregate(items []azdevops.WorkItem, intervalStart, now time.Time, th Thresholds) ([]UserMetrics, []ItemFlag) {
	byUser := make(map[string]*UserMetrics)
	get := func(u string) *UserMetrics {
		if byUser[u] == nil {
			byUser[u] = &UserMetrics{User: u}
		}
		return byUser[u]
	}

	activeStale := time.Duration(th.ActiveStaleDays) * 24 * time.Hour
	rftStale := time.Duration(th.RFTStaleDays) * 24 * time.Hour
	states := th.States

	var flags []ItemFlag

	for i := range items {
		wi := items[i]
		user := wi.AssignedToName()
		dwell := wi.TimeInCurrentState(now)
		s := wi.Fields.State

		switch {
		case states.IsActive(s):
			um := get(user)
			um.InFlight++
			um.ActiveCount++
			if dwell > um.OldestActive {
				um.OldestActive = dwell
			}
			if dwell > activeStale {
				um.Stalled++
				flags = append(flags, ItemFlag{
					ID:      wi.ID,
					Title:   wi.Fields.Title,
					Project: wi.ProjectDisplayName,
					User:    user,
					State:   wi.Fields.State,
					Dwell:   dwell,
					Reason:  reasonActiveStale,
				})
			}
		case states.IsRFT(s):
			um := get(user)
			um.InFlight++
			um.RFTCount++
			if dwell > um.OldestRFT {
				um.OldestRFT = dwell
			}
			if dwell > rftStale {
				um.Stalled++
				flags = append(flags, ItemFlag{
					ID:      wi.ID,
					Title:   wi.Fields.Title,
					Project: wi.ProjectDisplayName,
					User:    user,
					State:   wi.Fields.State,
					Dwell:   dwell,
					Reason:  reasonRFTStale,
				})
			}
		case states.IsClosed(s):
			// Item is in the configured Closed state; count its points only if
			// the close happened inside the interval window.
			if !wi.Fields.ClosedDate.IsZero() && wi.Fields.ClosedDate.After(intervalStart) {
				um := get(user)
				um.PointsClosed += wi.EffectivePoints()
			}
		}
	}

	rows := make([]UserMetrics, 0, len(byUser))
	for _, um := range byUser {
		um.Overloaded = um.InFlight > th.WIPLimit
		rows = append(rows, *um)
	}

	sort.Slice(rows, func(i, j int) bool {
		if rows[i].Stalled != rows[j].Stalled {
			return rows[i].Stalled > rows[j].Stalled
		}
		return rows[i].InFlight > rows[j].InFlight
	})
	sort.Slice(flags, func(i, j int) bool { return flags[i].Dwell > flags[j].Dwell })

	return rows, flags
}

File: src/Azdo.Core/Metrics/Aggregate.cs
using Azdo.Core.AzureDevOps;

namespace Azdo.Core.Metrics;

/// <summary>
/// Configurable cut-offs that drive flagging and the overloaded marker on the
/// per-user table.
/// </summary>
public sealed class Thresholds
{
    /// <summary>Dwell in Active longer than this → active-stale flag.</summary>
    public int ActiveStaleDays { get; set; }

    /// <summary>Dwell in Ready for Test longer than this → rft-stale flag.</summary>
    public int RFTStaleDays { get; set; }

    /// <summary>Strictly more in-flight items than this → Overloaded.</summary>
    public int WIPLimit { get; set; }

    /// <summary>Canonical state names; matched case-insensitively.</summary>
    public StateConfig States { get; set; } = StateConfig.DefaultStates();
}

/// <summary>One row of the per-developer table.</summary>
public sealed class UserMetrics
{
    public string User { get; set; } = "";
    public int InFlight { get; set; }
    public int ActiveCount { get; set; }
    public int RFTCount { get; set; }
    public double PointsClosed { get; set; }
    public TimeSpan OldestActive { get; set; }
    public TimeSpan OldestRFT { get; set; }
    public int Stalled { get; set; }
    public bool Overloaded { get; set; }
}

/// <summary>One row of the worst-first stuck-items digest.</summary>
public sealed class ItemFlag
{
    public int Id { get; set; }
    public string Title { get; set; } = "";
    public string Project { get; set; } = "";
    public string User { get; set; } = "";
    public string State { get; set; } = "";
    public TimeSpan Dwell { get; set; }

    /// <summary>"active-stale" or "rft-stale".</summary>
    public string Reason { get; set; } = "";
}

/// <summary>Pure aggregation logic for the metrics tab.</summary>
public static class Aggregator
{
    public const string ReasonActiveStale = "active-stale";
    public const string ReasonRFTStale = "rft-stale";

    /// <summary>
    /// Rolls a flat slice of work items into per-developer rows and a flat list of
    /// flagged items, applying <paramref name="th"/> thresholds. Items in states
    /// other than Active / Ready for Test / Closed are ignored. Closed items
    /// contribute to PointsClosed only when their ClosedDate falls strictly after
    /// <paramref name="intervalStart"/>.
    /// </summary>
    public static (List<UserMetrics> Rows, List<ItemFlag> Flags) Aggregate(
        IReadOnlyList<WorkItem> items, DateTime intervalStart, DateTime now, Thresholds th)
    {
        var byUser = new Dictionary<string, UserMetrics>();
        UserMetrics Get(string u)
        {
            if (!byUser.TryGetValue(u, out var um))
            {
                um = new UserMetrics { User = u };
                byUser[u] = um;
            }
            return um;
        }

        var activeStale = TimeSpan.FromDays(th.ActiveStaleDays);
        var rftStale = TimeSpan.FromDays(th.RFTStaleDays);
        var states = th.States;

        var flags = new List<ItemFlag>();

        foreach (var wi in items)
        {
            var user = wi.AssignedToName();
            var dwell = wi.TimeInCurrentState(now);
            var s = wi.Fields.State;

            if (states.IsActive(s))
            {
                var um = Get(user);
                um.InFlight++;
                um.ActiveCount++;
                if (dwell > um.OldestActive) um.OldestActive = dwell;
                if (dwell > activeStale)
                {
                    um.Stalled++;
                    flags.Add(new ItemFlag
                    {
                        Id = wi.Id,
                        Title = wi.Fields.Title,
                        Project = wi.ProjectDisplayName,
                        User = user,
                        State = wi.Fields.State,
                        Dwell = dwell,
                        Reason = ReasonActiveStale,
                    });
                }
            }
            else if (states.IsRFT(s))
            {
                var um = Get(user);
                um.InFlight++;
                um.RFTCount++;
                if (dwell > um.OldestRFT) um.OldestRFT = dwell;
                if (dwell > rftStale)
                {
                    um.Stalled++;
                    flags.Add(new ItemFlag
                    {
                        Id = wi.Id,
                        Title = wi.Fields.Title,
                        Project = wi.ProjectDisplayName,
                        User = user,
                        State = wi.Fields.State,
                        Dwell = dwell,
                        Reason = ReasonRFTStale,
                    });
                }
            }
            else if (states.IsClosed(s))
            {
                if (wi.Fields.ClosedDate != default && wi.Fields.ClosedDate > intervalStart)
                {
                    var um = Get(user);
                    um.PointsClosed += wi.EffectivePoints();
                }
            }
        }

        var rows = new List<UserMetrics>(byUser.Count);
        foreach (var um in byUser.Values)
        {
            um.Overloaded = um.InFlight > th.WIPLimit;
            rows.Add(um);
        }

        rows.Sort((a, b) =>
        {
            if (a.Stalled != b.Stalled) return b.Stalled.CompareTo(a.Stalled);
            return b.InFlight.CompareTo(a.InFlight);
        });
        flags.Sort((a, b) => b.Dwell.CompareTo(a.Dwell));

        return (rows, flags);
    }
}
